internal/controllers/machinedeployment/machinedeployment_controller.go

```diff
@@ -282,6 +295,33 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 	return errors.Errorf("unexpected deployment strategy type: %s", md.Spec.Strategy.Type)
 }
 
+func (r *Reconciler) reconcileDelete(ctx context.Context, md *clusterv1.MachineDeployment) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	msList, err := r.getMachineSetsForDeployment(ctx, md)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If all the descendant machinesets are deleted, then remove the machinedeployment's finalizer.
+	if len(msList) == 0 {
+		controllerutil.RemoveFinalizer(md, clusterv1.MachineDeploymentFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	// else delete owned machinesets.
+	log.Info("MachineDeployment still has owned MachineSets, deleting them first")
```

Comment:
In the Machine controller we log how many descendants are still in flight, could we do something similar here?

Reply:
Yes, please!

I think here we should also list the MachineSets by name that still remain (see cluster_controller.go "Cluster still has descendants ..." log line)

Reply:
@vincepri Did you mean in the Cluster controller?